File: table-engine/src/styles.ts
/**
 * Table Styles — Pure computation for table cell format resolution.
 *
 * Every function delegates to Rust/WASM via compute-core.
 * Built-in styles are cached after first retrieval since they never change.
 *
 * Contains all 67 built-in Excel table style definitions (28 Light, 28 Medium, 11 Dark).
 */

import type { Table, TableCellFormat, TableStyleDef } from './types';

import { getWasm } from './wasm-backend';

// =============================================================================
// Public API (delegates to WASM)
// =============================================================================

/**
 * Return all 67 built-in Excel table style definitions.
 * Cached after first retrieval since built-in styles never change.
 */
let cachedStyles: Record<string, TableStyleDef> | null = null;

export function getBuiltInTableStyles(): Record<string, TableStyleDef> {
  if (!cachedStyles) {
    cachedStyles = getWasm().table_get_built_in_styles() as Record<string, TableStyleDef>;
  }
  return cachedStyles;
}

/**
 * Resolve the cell format for a given grid position within a table.
 *
 * Pure function: Table + (row, col) -> TableCellFormat | null.
 *
 * Returns null if (row, col) is outside the table range.
 *
 * Resolution priority (highest to lowest):
 *   1. Header row
 *   2. Totals row
 *   3. First column emphasis
 *   4. Last column emphasis
 *   5. Column banding
 *   6. Row banding
 */
export function resolveTableCellFormat(
  table: Table,
  row: number,
  col: number,
): TableCellFormat | null {
  const result = getWasm().table_resolve_cell_format(table, row, col);
  return (result ?? null) as TableCellFormat | null;
}

File: compute/core/crates/compute-table/src/styles.rs
//! Table Styles - pure computation for table cell format resolution.
//!
//! Every function is pure and stateless. No DOM, no Yjs, no React.

mod borders;
mod builtins;
mod resolver;

use crate::types::{Table, TableCellFormat, TableStyleDef};

/// Default style used when the table's style ID is not found.
pub const DEFAULT_STYLE_ID: &str = "TableStyleMedium2";

/// Look up a built-in table style by ID.
pub fn get_built_in_style(id: &str) -> Option<&'static TableStyleDef> {
    builtins::get(id)
}

/// Return all built-in Excel table style definitions.
pub fn get_all_built_in_styles() -> Vec<&'static TableStyleDef> {
    builtins::all()
}

/// Resolve the cell format for a given grid position within a table.
pub fn resolve_table_cell_format(table: &Table, row: u32, col: u32) -> Option<TableCellFormat> {
    resolver::resolve_table_cell_format(table, row, col)
}
